### PR TITLE
pkg/logger performance improvements

### PR DESCRIPTION
**Pull request by aanm**
```release-note
Improved performance of endpoint logging
```

Please check per commit basis

---
This change is [Reviewable](https://reviewable.io/reviews/cilium/cilium/5270)

### CODE REVIEW

**Comment by aanm** [comment]
test-me-please

**Comment by aanm** [comment]
build failed while downloading new VM re-triggering

**Comment by aanm** [comment]
test-me-please

File: pkg/endpoint/log.go [hunk 1]
```diff
@@ -75,16 +78,16 @@ func (e *Endpoint) updateLogger() {
 		baseLogger.SetLevel(logrus.DebugLevel)
 	}
 
-	e.loggerMutex.Lock()
-	// Note: endpoint.loggerMutex protects the reference but not the logger objects. We
-	// cannot update the old object directly as that could be racey.
-	e.logger = baseLogger.WithFields(logrus.Fields{
+	l := baseLogger.WithFields(logrus.Fields{
```
Comment (tgraf) [inline pkg/endpoint/log.go]: Don't we need to serialize calls to `updateLogger()`? Does it matter?

**Comment by aanm** [comment]
Build failed, https://jenkins.cilium.io/job/Cilium-PR-Ginkgo-Tests-Validated/5767/ triggering a new build to discard anything related with my changes.

**Comment by aanm** [comment]
test-me-please

File: pkg/endpoint/log.go (hunk 1 above)
Comment (aanm) [reply pkg/endpoint/log.go]: I don't think it matters TBH

**Comment by aanm** [comment]
test-me-please

**Review by ianvernon — APPROVED** [review]

File: daemon/endpoint.go [hunk 2]
```diff
@@ -436,6 +436,7 @@ func (h *patchEndpointID) Handle(params PatchEndpointIDParams) middleware.Respon
 		}
 	}
 
+	ep.UpdateLogger(nil)
```
Comment (tgraf) [inline daemon/endpoint.go:439]: I feel a bit uneasy that we will forget to call this in places. How do developers know when to call this?

**Review by tgraf — CHANGES_REQUESTED** [review]

File: pkg/endpoint/log.go (hunk 1 above)
Comment (aanm) [reply pkg/endpoint/log.go]: @tgraf here

File: daemon/endpoint.go (hunk 2 above)
Comment (aanm) [reply daemon/endpoint.go:439]: @tgraf 
The `UpdateLogger` is used on all Seters that exist for the endpoint and in there, the UpdateLogger is already being called. Would it help if I add a comment where the logger is being created with the fields?

File: pkg/endpoint/log.go [hunk 3]
```diff
@@ -29,11 +32,9 @@ var log = logging.DefaultLogger.WithField(logfields.LogSubsys, "endpoint")
 func (e *Endpoint) getLogger() *logrus.Entry {
 	e.updateLogger()
 
-	e.loggerMutex.RLock()
```
Comment (rlenglet) [inline pkg/endpoint/log.go]: Remove the `loggerMutex` field from `Endpoint`. It's not used anywhere anymore.

File: pkg/endpoint/log.go [hunk 4]
```diff
@@ -75,16 +78,16 @@ func (e *Endpoint) updateLogger() {
...
 		logfields.EndpointID:     e.ID,
 		logfields.ContainerID:    containerID,
 		logfields.PolicyRevision: e.policyRevision,
 		logfields.IPv4:           e.IPv4.String(),
 		logfields.IPv6:           e.IPv6.String(),
 		logfields.K8sPodName:     podName,
 	})
-	e.loggerMutex.Unlock()
+
+	// Note: endpoint.loggerMutex protects the reference but not the logger objects. We
```
Comment (rlenglet) [inline pkg/endpoint/log.go]: Remove this comment. `loggerMutex` is not used anymore.

File: pkg/endpoint/endpoint.go [hunk 5]
```diff
@@ -491,19 +491,13 @@ func (e *Endpoint) WaitForProxyCompletions(proxyWaitGroup *completion.WaitGroup)
 
 	start := time.Now()
 
-	if err := e.RLockAlive(); err != nil {
```
Comment (rlenglet) [inline pkg/endpoint/endpoint.go:494]: This is great!

**Review by rlenglet — APPROVED** [review]
This is great. Just a few nitpicks.

Also, please replace the `logger` local variable in `Endpoint.regenerateBPF` with calls to `e.getLogger()`. The state of the endpoint often changes during regeneration, and using a cached logger makes the logs very confusing when debugging this.

Same for other regeneration methods: `Endpoint.regenerate`, `Endpoint.regeneratePolicy`, `Endpoint.UpdateLabels` and the methods it calls, etc.

File: daemon/endpoint.go (hunk 2 above)
Comment (tgraf) [reply daemon/endpoint.go:439]: Adding a comment to all fields in the endpoint struct to make sure we don't write to them directly is enough for me. Good to see that this is already abstracted into setters.

**Review by tgraf — APPROVED** [review]

**Comment by tgraf** [comment]
Approving assuming Romain's feedback gets fixed and a comment is added.

File: pkg/endpoint/policy.go [hunk 6]
```diff
@@ -693,11 +693,7 @@ func (e *Endpoint) regenerate(owner Owner, reason string) (retErr error) {
 	e.BuildMutex.Lock()
 	defer e.BuildMutex.Unlock()
 
-	if err = e.RLockAlive(); err != nil {
```
Comment (tgraf) [inline pkg/endpoint/policy.go]: Actually, we need to check whether the endpoint is alive after taking the BuildMutex. We don't want to build dead endpoints.

**Review by tgraf — CHANGES_REQUESTED** [review]

**Comment by aanm** [comment]
test-me-please

**Review by tgraf — APPROVED** [review]

**Comment by aanm** [comment]
test-me-please

**Comment by tgraf** [comment]
Lots of future code will depend on this assumption and it will be extremely hard to always remember this while reviewing so I suggest we backport this.

**Pull request merged**